File: algorithms/Reverse Subarray To Maximize Array Value/Reverse Subarray To Maximize Array Value.py
class Solution(object):
    def maxValueAfterReverse(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # basic idea without fancy stuff  
        # 1. https://leetcode.com/problems/reverse-subarray-to-maximize-array-value/discuss/489929/O(n)-time-O(1)-space.-In-depth-Explanation
        # 2. https://code.dennyzhang.com/reverse-subarray-to-maximize-array-value
        
        # This can be done in three steps each step taking O(N) time
        
        res, n = 0, len(nums)
        
        # Step 1: Assume that no subarray is reversed and so the sum would just accumulate over all the abs differences
        for index in range(1, n):
            res += abs(nums[index] - nums[index - 1])
        
        # Step 2: Reversing the left or the right half so basically this idea stems from prefix array type question where in which we have the sum upto ith index but then to get at a specific index we substract extra sum, following from that idea. Or refer to the reference 1
        
        diff = 0
        for index in range(n):
            
            # reversing from 0th index to current
            if index + 1 < n:
                diff = max(diff, abs(nums[index + 1] - nums[0]) - abs(nums[index + 1] - nums[index])) # diff between 0 and curr - diff curr and curr + 1
            
            # reversing from current to last index n - 1
            if index > 0:
                diff = max(diff, abs(nums[index - 1] - nums[n - 1]) - abs(nums[index - 1] - nums[index]))
                
        
        # Step 3: We still need to check the middle reverse part, we can do this using the min max trick
        low_number, high_number = float("inf"), float("-inf")
        
        for index in range(n - 1):
            
            low_number = min(low_number, max(nums[index], nums[index + 1])) # min of low and the max of the current and next number
            high_number = max(high_number, min(nums[index], nums[index + 1]))
        
			diff = max(diff, 2 * (high_number - low_number)) # This is explained in ref 1
        
        return res + diff
            
        

File: algorithms/R/Reverse Subarray To Maximize Array Value/Reverse Subarray To Maximize Array Value.cpp
class Solution {
public:

int maxValueAfterReverse(vector<int>& nums) {        
    int original =  0;        
    for(int i = 0 ;  i < nums.size() - 1 ; ++i) {
           original += abs(nums[i] - nums[i+1]);
    }       
    int ans  =  original;  
    int a1 = INT_MAX; 
    int a2 = INT_MAX;  
    int a3 = INT_MIN;   
    int a4 = INT_MIN; 
    int n = nums.size();
    for(int i = n - 2; i >= 1; --i) {
            int old = nums[i-1];
            int now = nums[i]; 
            if(i != n-2) {   
                int y1 = (nums[i-1] + nums[i]) - abs(old - now) - a1 + original;
                int y2 = (nums[i-1] - nums[i]) - abs(old - now) - a2 + original;
                int y3 = (nums[i] - nums[i-1]) - abs(old - now) + a3 + original;
                int y4 = -(nums[i-1] + nums[i]) - abs(old - now) + a4 + original;   
                ans =  max(max(y4, max(y1, max(y2,y3))), ans);  
            }
            a1 = min(a1, nums[i] + nums[i+1] + abs(nums[i] - nums[i+1])); 
            a2 = min(a2, nums[i] - nums[i+1] + abs(nums[i] - nums[i+1])); 
            a3 = max(a3, nums[i] - nums[i+1] - abs(nums[i] - nums[i+1])); 
            a4 = max(a4, nums[i] + nums[i+1] - abs(nums[i] - nums[i+1]));
            ans = max(ans, -abs(old - now) + abs(old - nums[nums.size()-1] + original)); 
    }
    for(int i = 1; i < n- 1; ++i) {
        ans = max(ans, -abs(nums[i] -nums[i+1]) + abs(nums[0] - nums[i + 1]) + original); 
    }
    return ans; 
}
